core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -55,23 +55,23 @@ object KafkaRequestHandler {
    * @param fun Callback function to execute
    * @return Wrapped callback that would execute `fun` on a request thread
    */
-  def wrap[T](fun: T => Unit): T => Unit = {
+  def wrap[T](fun: (RequestLocal, T) => Unit, requestLocal: RequestLocal): T => Unit = {
     val requestChannel = threadRequestChannel.get()
     val currentRequest = threadCurrentRequest.get()
     if (requestChannel == null || currentRequest == null) {
       if (!bypassThreadCheck)
         throw new IllegalStateException("Attempted to reschedule to request handler thread from non-request handler thread.")
-      T => fun(T)
+      T => fun(requestLocal, T)
     } else {
       T => {
-        if (threadCurrentRequest.get() != null) {
-          // If the callback is actually executed on a request thread, we can directly execute
+        if (threadCurrentRequest.get() == currentRequest) {
+          // If the callback is actually executed on the same request thread, we can directly execute
```

Comment:
Without this, there were a ton of failing tests when David implemented his change. We never really figured out the issue. We also have to consider the case where we bypass the thread check (ie tests) where we don't have the callback queue set up.

Reply:
@jolshan : I wasn't suggesting simply skip the test `threadCurrentRequest.get() == currentRequest`. I was wondering if we could add the callback to the `callbackQueue` through the following call, independent of whether the caller is in the same request thread of not. 

` requestChannel.sendCallbackRequest`

Reply:
Right -- we had that before and it caused issues @dajac has more context. 
In addition, for tests, we would need to set up the request channel etc.

Reply:
Thanks, Justine. It seems that we can't easily get rid of the ThreadLocal stuff. So, we can leave this as it is.

> I am concerned that executing the callback and before we return from the request thread causes issues.

With the current PR, it seems that it's possible for the callback to be called before the request handler thread finishes processing the request (that generates the callback), right? Is that causing any problem?

Reply:
That seemed to cause problems when David moved changed the code to do the callback before we finished processing the request. 

We never figured out the root cause though.

Reply:
So, is this still a problem that we need to address in this PR?

Reply:
Sorry maybe I'm not explaining myself well.

We saw problems before when the callback was scheduled before the request returned. We fixed it by adding this check.

If we want to remove this check, we will need to do further investigation for that case.

Reply:
My understanding is that concurrency is not a problem in production logic (i.e. the fact that callback might actually execute at any time before or after or concurrently with the passing thread), but unit tests rely on deterministic order of execution and so adding concurrency here makes them "flaky".
I agree in principle that if we could fold this logic into just one case (i.e. always schedule on a new request thread), it would make it simpler; when this change was proposed I looked into unit test fix and seemed more involved than adding this logic.

Reply:
Thanks for the explanation, Justine and Artem. This sounds good then.

Reply:
Can we file a JIRA for investigating the alternative? I agree with @junrao that it would be much simpler to reason about if we always schedule it.

Reply:
https://issues.apache.org/jira/browse/KAFKA-15758